cirq/experiments/random_quantum_circuit_generation_test.py

```diff
@@ -148,6 +151,87 @@ def test_get_random_combinations_for_small_device():
     assert len(combinations) == 2  # 3x1 device only fits two layers
 
 
+def test_get_random_combinations_for_pairs():
+    all_pairs = [
+        [(cirq.LineQubit(0), cirq.LineQubit(1)), (cirq.LineQubit(2), cirq.LineQubit(3))],
+        [(cirq.LineQubit(1), cirq.LineQubit(2))],
+    ]
+    combinations = get_random_combinations_for_pairs(
+        n_library_circuits=3, n_combinations=4, all_pairs=all_pairs, random_state=99
+    )
+    assert len(combinations) == len(all_pairs)
+    for i, comb in enumerate(combinations):
+        assert comb.combinations.shape[0] == 4  # n_combinations
+        assert comb.combinations.shape[1] == len(comb.pairs)
+        assert np.all(comb.combinations >= 0)
+        assert np.all(comb.combinations < 3)  # number of library circuits
+        for q0, q1 in comb.pairs:
+            assert q0 in cirq.LineQubit.range(4)
+            assert q1 in cirq.LineQubit.range(4)
+
+        assert comb.layer is None
+        assert comb.pairs == all_pairs[i]
+
+
+def test_get_random_combinations_for_layer_circuit():
+    q0, q1, q2, q3 = cirq.LineQubit.range(4)
+    circuit = cirq.Circuit(cirq.CNOT(q0, q1), cirq.CNOT(q2, q3), cirq.CNOT(q1, q2))
+    combinations = get_random_combinations_for_layer_circuit(
+        n_library_circuits=3, n_combinations=4, layer_circuit=circuit, random_state=99
+    )
+    assert len(combinations) == 2  # operations pack into two layers
+    for i, comb in enumerate(combinations):
+        assert comb.combinations.shape[0] == 4  # n_combinations
+        assert comb.combinations.shape[1] == len(comb.pairs)
+        assert np.all(comb.combinations >= 0)
+        assert np.all(comb.combinations < 3)  # number of library circuits
+        for q0, q1 in comb.pairs:
+            assert q0 in cirq.LineQubit.range(4)
+            assert q1 in cirq.LineQubit.range(4)
+
+        assert comb.layer == circuit.moments[i]
+
+
+def test_get_random_combinations_for_bad_layer_circuit():
+    q0, q1, q2, q3 = cirq.LineQubit.range(4)
+    circuit = cirq.Circuit(
+        cirq.H.on_each(q0, q1, q2, q3), cirq.CNOT(q0, q1), cirq.CNOT(q2, q3), cirq.CNOT(q1, q2)
+    )
+
+    with pytest.raises(ValueError, match=r'>2-qubit operation'):
+        _ = get_random_combinations_for_layer_circuit(
+            n_library_circuits=3, n_combinations=4, layer_circuit=circuit, random_state=99
+        )
+
+
+def test_get_grid_interaction_layer_circuit():
+    graph = _gridqubits_to_graph_device(cirq.GridQubit.rect(3, 3))
+    layer_circuit = get_grid_interaction_layer_circuit(graph)
+    cirq.testing.assert_has_diagram(
```

Comment:
Out of curiosity, isn't it better to compare against a manually constructed expected circuit?

Reply:
yeah, you're right. I put in a manually constructed one with comment annotations about why we'd expect it to be that way

Reply:
I also added a check that get_random_combinations_for_layer_circuit gives the same result as get_random_combinations_for_device when using get_grid_interaction_layer_circuit